Update network state on error condition.

asurdej-comcast opened the pull request:
Set the NETWORK_IDLE state on decode or network errors for high enough ready state
as specified in '4.8.11.5 Loading the media resource':
https://html.spec.whatwg.org/multipage/media.html#loading-the-media-resource

This change fixes orphant HTML video tag removal that is in error state.
removedFromAncestor() -> pauseAfterDetachedTask() cycle
will trigger pause() switching m_playing to 0 so the object
won't report any pending activity from virtualHasPendingActivity()
and can be cleared by GC.

Commit:
Set media elements' networkState to NETWORK_IDLE if the connection is interrupted

This is a backport of https://bugs.webkit.org/show_bug.cgi?id=223242

Patch by Souju TANAKA <soujyu.tanaka@access-company.com> on 2021-03-17
Reviewed by Eric Carlson.

Media elements' networkState was set to NETWORK_EMPTY if the connection is interrupted after some
media data has been received. This fix changes the state to NETWORK_IDLE, according to recent spec.

No new tests since there seems no way to disconnect network connection during tests.

* html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaLoadingFailedFatally):

Canonical link: https://commits.webkit.org/235404@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@274559 268f45cc-cd09-0410-ab3c-d52691b4dbfc

asurdej-comcast commented:
@eocanha Would you mind taking a look at this patch? The exact piece of spec for this change is:
https://html.spec.whatwg.org/multipage/media.html#loading-the-media-resource
"
If the [media data](https://html.spec.whatwg.org/multipage/media.html#media-data) is corrupted

    Fatal errors in decoding the [media data](https://html.spec.whatwg.org/multipage/media.html#media-data) that occur after the user agent has established whether the current media resource is usable (i.e. once the [media element](https://html.spec.whatwg.org/multipage/media.html#media-element)'s [readyState](https://html.spec.whatwg.org/multipage/media.html#dom-media-readystate) attribute is no longer [HAVE_NOTHING](https://html.spec.whatwg.org/multipage/media.html#dom-media-have_nothing)) must cause the user agent to execute the following steps:

        The user agent should cancel the fetching process.

        Set the [error](https://html.spec.whatwg.org/multipage/media.html#dom-media-error) attribute to the result of [creating a MediaError](https://html.spec.whatwg.org/multipage/media.html#creating-a-mediaerror) with [MEDIA_ERR_DECODE](https://html.spec.whatwg.org/multipage/media.html#dom-mediaerror-media_err_decode).

        Set the element's [networkState](https://html.spec.whatwg.org/multipage/media.html#dom-media-networkstate) attribute to the [NETWORK_IDLE](https://html.spec.whatwg.org/multipage/media.html#dom-media-network_idle) value.

        Set the element's [delaying-the-load-event flag](https://html.spec.whatwg.org/multipage/media.html#delaying-the-load-event-flag) to false. This stops [delaying the load event](https://html.spec.whatwg.org/multipage/parsing.html#delay-the-load-event).

        [Fire an event](https://dom.spec.whatwg.org/#concept-event-fire) named [error](https://html.spec.whatwg.org/multipage/media.html#event-media-error) at the [media element](https://html.spec.whatwg.org/multipage/media.html#media-element).

        Abort the overall [resource selection algorithm](https://html.spec.whatwg.org/multipage/media.html#concept-media-load-algorithm).

asurdej-comcast commented:
There is an upstream change already that does pretty much the same:
https://bugs.webkit.org/show_bug.cgi?id=223242
I will do the backport instead of creating new one

calvaris commented:
> There is an upstream change already that does pretty much the same: https://bugs.webkit.org/show_bug.cgi?id=223242 I will do the backport instead of creating new one

I had filed https://bugs.webkit.org/show_bug.cgi?id=223242 and its corresponding PR but I closed it as Resolved Later and marked the PR as draft while waiting for this.

asurdej-comcast committed:
[HTMLMediaPlayer] Destroy media player on fatal playback error

Clean up media player on fata error so it can't be reused on further
playback attempts.

This fixes a crash when media player tries to reuse a media source
that was detached from media element already due to error.

asurdej-comcast commented:
Hi @calvaris I created a backport of the upstream change for setting NETWORK_IDLE and for detachingMediaSource() I made a step ahead and add a change that will destroy the player completely:
Moved all of this to new PR/review: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/899
Could you please take a look?

I'm closing this PR as all the changes were moved to #899
